tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp

```diff
@@ -0,0 +1,532 @@
+} // namespace indirect_unary_predicate_test
+
+namespace indirect_binary_predicate_test {
+    // Also validate indirect_result_t
+    using std::indirect_binary_predicate;
+
+    template <int>
+    struct base {};
+    template <>
+    struct base<0> {
+        // 0: not copy_constructible
+        base()            = default;
+        base(base const&) = delete;
+    };
+
+    template <int I>
+    struct Fn : base<I> {
+        // 1: not predicate<Fn&, iter_value_t<simple_iter_archetype>&, iter_value_t<simple_iter_archetype>&>
+        void operator()(int&, int&) const requires(I == 1);
+        // 2: not predicate<Fn&, iter_value_t<simple_iter_archetype>&, iter_reference_t<simple_iter_archetype>>
+        void operator()(int&, simple_reference) const requires(I == 2);
+        // 3: not predicate<Fn&, iter_reference_t<simple_iter_archetype>, iter_value_t<simple_iter_archetype>&>
+        void operator()(simple_reference, int&) const requires(I == 3);
+        // 4: not predicate<Fn&, iter_reference_t<simple_iter_archetype>, iter_reference_t<simple_iter_archetype>>
+        void operator()(simple_reference, simple_reference) const requires(I == 4);
+        // 5: not predicate<Fn&, iter_common_reference_t</**/>, iter_common_reference_t</**/>>
+        void operator()(simple_common_reference, simple_common_reference) const requires(I == 5);
+
+        bool operator()(int&, int&) const requires(I != 1);
+        int operator()(int&, simple_reference) const requires(I != 2);
+        int* operator()(simple_reference, int&) const requires(I != 3);
+        std::true_type operator()(simple_reference, simple_reference) const requires(I != 4);
+        std::false_type operator()(simple_common_reference, simple_common_reference) const requires(I != 5);
+    };
+
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<0>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<1>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<2>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<3>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<4>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<5>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<6>, simple_iter_archetype<0>, simple_iter_archetype<1>>);
+    STATIC_ASSERT(!indirect_binary_predicate<Fn<6>, simple_iter_archetype<1>, simple_iter_archetype<0>>);
+    STATIC_ASSERT(indirect_binary_predicate<Fn<6>, simple_iter_archetype<1>, simple_iter_archetype<1>>);
+
+    STATIC_ASSERT(std::same_as<std::indirect_result_t<Fn<6>, simple_iter_archetype<1>, simple_iter_archetype<1>>,
+        std::true_type>);
+} // namespace indirect_binary_predicate_test
+
+// TODO: test indirect_equivalence_relation

```

Comment:
Given the "hot" nature of this change I won't go "over my dead body" over checking in TODOs, but there should be GH issues for it

Reply:
Gack, I need to make another pass to remove these unused/untested concepts. (I pasted in a bunch of untested support machinery from another branch, put TODOs here, and added test coverage for bits that I needed for algorithms as I went. I had intended to remove these before pushing.)